### ISSUE TITLE
Xref confusingly jumps to temp-file copies of buffer

### ISSUE THREAD

**Issue by shajra** [jyp/dante#78]
When I have a module that references another with the following files (Cabal file not shown, but what you would expect)

**src/Foo.hs:**
```haskell
module Foo where
import Bar (myNum)
nextNum = myNum + 1
```

**src/Bar.hs:**
```haskell
module Bar where
myNum = 42
```

**~/.emacs/init.el**
```elisp
(require 'package)
(setq package-enable-at-startup nil)
(add-to-list 'package-archives '("melpa" . "http://melpa.org/packages/"))
(package-initialize)
(unless (package-installed-p 'use-package)
  (package-refresh-contents)
  (package-install 'use-package))

(use-package flycheck
  :ensure t
  :init (global-flycheck-mode))

(use-package dante
  :ensure t
  :after haskell-mode
  :commands 'dante-mode
  :init
  (add-hook 'haskell-mode-hook 'dante-mode)
  (add-hook 'haskell-mode-hook 'flycheck-mode))
```
I can point at the `myNum` reference in the "Foo.hs" buffer and jump with `xref-find-definitions` to a buffer for the sibling Bar.hs file.  But if I go back to the "Foo.hs" buffer (not closing the "Bar.hs" buffer) and try to jump again, I get sent to the "dante-temp-file-name" buffer for Bar.hs, which for me is something like /run/user/$UID/dante3693Ygk.hs.

This is pretty inconvenient because the temporary file looks identical to the real source file, and it's easy to accidentally start editing its buffer instead of the real one.

I tried to understand more of the Dante code, but didn't get far enough to figure out what exactly is going on and where a fix might be.  I'm hoping we don't get stuck at "works as designed" because of the way temp-files are managed by Dante.

I also tried to reproduce this problem without Flycheck, but couldn't figure out a way.  So I suspect this is an interaction between Xref integration and Flycheck.

My Emacs configuration and the project are fairly minimized, so I'm hoping this is reproducible.

**Comment by emilypi** [comment]
I've been experiencing something similar with respect to `xref-find-definition` jumping to a tags file in `/var/folder/.../`. `C-c ,` is also jumping to this file for info hints, and reporting the incorrect file location. Perhaps these are related? You can see it [here](https://i.imgur.com/RevjPsg.jpg).

**Comment by jyp** [comment]
The problem is that GHCi associates the temporary file with the module. I am thinking that the best method would be to stop using temporary files, and simply save the file before calling into ghci.

**Comment by shajra** [comment]
This should work fine with setting `flycheck-check-syntax-automatically` to  `'(mode-enabled save)`?  I'm not saying it shouldn't.  I just don't know enough about interactions between packages.

**Comment by jyp** [comment]
@shajra I've implemented something already in the `no-temps` branch. Please test if it works. I did not know about `'(mode-enabled save)`, so I did the saves manually in dante.

**Comment by jyp** [comment]
Ok, I see that it's a customize variable, so I suppose that I should not fiddle with it anyway.

**Comment by jyp** [comment]
dante no longer creates temp files.

**Comment by emilypi** [comment]
Thanks @jyp !

**Comment by jyp** [comment]
@sboosali Dante will save the current buffer every time that a check is initiated, regardless of the customization.

**Comment by soupi** [comment]
Sorry, but isn't that a bit extreme? Saving the buffer for the user any time they change it can trigger all kind of unwanted things (`stack build --file-watch`, `ghcid`, makefiles rebuild because time was modified). Can this be turned off somehow?

**Comment by jyp** [comment]
@soupi The way to go is to use `'(mode-enabled save)`, like @shajra was suggesting. I'd like to be able to configure this from within dante but I'm not sure how to do it.

**Comment by purcell** [comment]
> Saving the buffer for the user any time they change it can trigger all kind of unwanted things

If you're interested in testing my fix for this, @soupi, check out #122 and let me know there if you see any problems.
